R/parse-body.R

```diff
@@ -16,230 +12,378 @@ postBodyFilter <- function(req){
   forward()
 }
 
-parseBody <- function(body, content_type = NULL) {
+postbody_parser <- function(req, parsers = NULL) {
+  if (length(parsers) == 0) {return(list())}
+  type <- req$HTTP_CONTENT_TYPE
+  body <- req$postBodyRaw
+  if (is.null(body)) {return(list())}
+  parse_body(body, type, parsers)
+}
+
+parse_body <- function(body, content_type = NULL, parsers = NULL) {
   if (!is.raw(body)) {body <- charToRaw(body)}
-  toparse <- list(value = body, content_type = content_type)
-  parseRaw(toparse)
+  toparse <- list(value = body, content_type = content_type, parsers = parsers)
+  parse_raw(toparse)
 }
 
-parseRaw <- function(toparse) {
+parse_raw <- function(toparse) {
   if (length(toparse$value) == 0L) return(list())
-  parser <- parserPicker(toparse$content_type, toparse$value[1], toparse$filename)
-  do.call(parser(), toparse)
+  parser <- parser_picker(
+    # Lower case content_type for parser matching
+    tolower(toparse$content_type),
+    toparse$value[1],
+    toparse$filename,
+    toparse$parsers
+  )
+  if (is.null(parser)) {
+    message("No suitable parser found to handle request body type ", toparse$content_type, ".")
+    return(list())
+  }
+  do.call(parser, toparse)
 }
 
-parserPicker <- function(content_type, first_byte, filename = NULL) {
-  #fast default to json when first byte is 7b (ascii {)
-  if (first_byte == as.raw(123L)) {
-    return(.globals$parsers$func[["json"]])
+parser_picker <- function(content_type, first_byte, filename = NULL, parsers = NULL) {
+
+  # parse as a query string
+  if (length(content_type) == 0) {
+    # fast default to json when first byte is 7b (ascii {)
+    if (first_byte == as.raw(123L)) {
+      return(parsers$json)
+    }
+
+    return(parsers$query)
   }
-  if (is.null(content_type)) {
-    return(.globals$parsers$func[["query"]])
+
+  # remove trailing content type information
+  # "application/json; charset=UTF-8"
+  # to
+  # "application/json"
+  if (stri_detect_fixed(content_type, ";")) {
+    content_type <- stri_split_fixed(content_type, ";")[[1]][1]
   }
-  # else try to find a match
-  patterns <- .globals$parsers$pattern
-  parser <- .globals$parsers$func[stri_startswith_fixed(content_type, patterns)]
-  # Should we warn when multiple parsers match?
-  # warning("Multiple body parsers matches for content-type : ", toparse$content_type, ". Parser ", names(parser)[1L], " used.")
-  if (length(parser) == 0L) {
-    if (is.null(filename)) {
-      return(.globals$parsers$func[["query"]])
-    } else {
-      return(.globals$parsers$func[["octet"]])
-    }
-  } else {
-    return(parser[[1L]])
+
+  parser <- parsers$fixed[[content_type]]
+
+  # return known parser (exact match)
+  if (!is.null(parser)) {
+    return(parser)
   }
-}
 
+  fpm <- stri_detect_regex(
+    content_type,
+    names(parsers$regex),
+    max_count = 1)
+  fpm[is.na(fpm)] <- FALSE
```

Comment:
Both `any()` and `which()` are `NA` safe

```suggestion
  fpm <- stri_detect_regex(
    content_type,
    names(parsers$regex),
    max_count = 1
  )
```